### PR TITLE
feat: Cohort - Retrieve, Replay & Audit

### PR DESCRIPTION
**Pull request by Warren-Pitterson**
## Context

Cohort - Retrieve, Replay & Audit functionality 

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ x] I have followed the code style of the project
- [ x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/RetrieveCohortDistribution.cs [hunk 1]
```diff
@@ -9,40 +9,51 @@ namespace NHS.CohortManager.CohortDistributionDataServices;
...
-        _httpRequestHelper = httpRequestHelper;
+        _httpParserHelper = httpParserHelper;
     }
 
     [Function("RetrieveCohortDistributionData")]
     public async Task<HttpResponseData> RunAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req)
     {
-        int serviceProviderId = HttpRequestHelper.GetServiceProviderId(req);
-        int rowCount = HttpRequestHelper.GetRowCount(req);
+        int screeningServiceId = _httpParserHelper.GetScreeningServiceId(req);
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/RetrieveCohortDistribution.cs:44]: I think there needs to be some error handling around these to catch bad requests if these arent passed as arguments

File: application/CohortManager/src/Functions/Shared/Data/Database/CreateCohortDistributionData.cs [hunk 2]
```diff
@@ -465,6 +500,7 @@ private bool UpdateRecords(List<SQLReturnModel> sqlToExecute)
             transaction.Rollback();
             _dbConnection.Close();
             _logger.LogError(ex, "An error occurred while inserting new Cohort Distribution records: {ExceptionMessage}", ex.Message);
+            _logger.LogError(ex, "An error occurred while inserting new Cohort Distribution records: {ExceptionMessage}", ex.Message);
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/CreateCohortDistributionData.cs:503]: It looks like we have a duplicate log statement.

File: application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/Program.cs [hunk 3]
```diff
@@ -21,6 +21,7 @@
         services.AddSingleton<ICreateCohortDistributionData, CreateCohortDistributionData>();
```
[the remark this answers is not in the document]
Comment (Warren-Pitterson) [reply application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/Program.cs]: Changed service intstance type to Transient

File: application/CohortManager/src/Functions/Shared/Data/Database/CreateCohortDistributionData.cs (hunk 2 above)
Comment (Warren-Pitterson) [reply application/CohortManager/src/Functions/Shared/Data/Database/CreateCohortDistributionData.cs]: removed duplicate error message

File: application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/RetrieveCohortDistribution.cs (hunk 1 above)
Comment (Warren-Pitterson) [reply application/CohortManager/src/Functions/CohortDistributionServices/RetrieveCohortDistribution/RetrieveCohortDistribution.cs:44]: Added error handling

File: application/CohortManager/src/Functions/Shared/Data/Database/IHttpParserHelper.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/IHttpParserHelper.cs:1]: Why is this in the database namespace?

Comment (Warren-Pitterson) [reply application/CohortManager/src/Functions/Shared/Data/Database/IHttpParserHelper.cs:1]: Moved to the shared interface folder and updated references

**Review by MWClayson-NHS — APPROVED** [review]

**Review by andrewmorris43 — APPROVED** [review]

**Pull request merged**